### PR TITLE
Add clip model and lightning

### PR DESCRIPTION
**Pull request by tourniert**
- Adding CLIP model (torch.nn.Module)
- Adding LightningModule to train a CLIP model
- Fix MiniTokenizer instanciation

### CODE REVIEW

File: mfai/tokenizers.py [hunk 1]
```diff
@@ -117,19 +117,24 @@ class MiniTokenizer(Tokenizer):
...
-        self.id_to_token: dict[int, int] | None = None
+        self.token_to_id: dict[int, int] = dict()
+        self.id_to_token: dict[int, int] = dict()
+
+        self.post_init(tokens)
+
+    @abstractmethod
+    def get_list_tokens(self) -> set:
+        """
+        Method that return a set of words to tokenize.
```
Comment (LBerth) [inline mfai/tokenizers.py:130]: I feel like the method name and the comment don't match : do we return words (str) or tokens (int) ? Maybe add that this will constitute the vocabulary of the tokenizer

File: mfai/tokenizers.py [hunk 2]
```diff
@@ -144,28 +149,28 @@ def name(self) -> str:
 
     def encode(self, text: str, *args: Any, **kwargs: Any) -> List[int]:
         base_token_ids = self.base_tokenizer.encode(text)
-        if self.token_to_id is not None:
-            return [self.token_to_id[x] for x in base_token_ids]
-        else:
-            return base_token_ids
+        # if self.token_to_id is not None:
+        #     return [self.token_to_id[x] for x in base_token_ids]
+        # else:
```
Comment (LBerth) [inline mfai/tokenizers.py:154]: remove commented lines here and below ?

File: mfai/torch/lightning_modules/clip.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
+from typing import Literal, Tuple, Union
```
Comment (LBerth) [inline mfai/torch/lightning_modules/clip.py:1]: add docstring that describes what CLIP does and link to paper ?

File: mfai/torch/lightning_modules/clip.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        """
+        Lightning method to define optimizers and learning-rate schedulers used for optimization.
+        For more details about this method, please see:
+        https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.core.LightningModule.html#lightning.pytorch.core.LightningModule.configure_optimizers
+        """
+        optimizer = AdamW(self.parameters(), lr=self.learning_rate)
+        if self.lr_scheduler_interval in ["step", "epoch"]:
+            num_batches = len(self.trainer.datamodule.train_dataloader())
+            warmup_epochs = num_batches if self.lr_scheduler_interval == "step" else 1
+            if self.trainer.max_steps > 0:
```
Comment (LBerth) [inline mfai/torch/lightning_modules/clip.py:68]: Why do you need to define this function ? Giving the class path and arguments for the optimizer and scheduler in the yaml config file doesn't work ?

File: mfai/torch/models/clip.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
+from dataclasses import dataclass
+from typing import Tuple, Union
+
+import torch
+
+# import torch.nn.functional as F
```
Comment (LBerth) [inline mfai/torch/models/clip.py:6]: rm commented lines

File: mfai/torch/lightning_modules/clip.py
Comment (LBerth) [inline mfai/torch/lightning_modules/clip.py:1]: Would it be possible to provide an example script to train CLIP on a famous dataset ? like we did with the Oxford pets and the segmentation module in the `examples` folder ? or if its too complicated, with a dummy dataset adapted to the task ?

File: mfai/torch/lightning_modules/clip.py (hunk 4 above)
Comment (tourniert) [reply mfai/torch/lightning_modules/clip.py:68]: I need the `configure_optimizers` method because the number of warmup steps and the total steps is defined in function of the scheduler interval ("step" or "epoch").

File: mfai/torch/lightning_modules/clip.py
Comment (tourniert) [reply mfai/torch/lightning_modules/clip.py:1]: Since LightningModule expects a `NamedTensor` for visual inputs, it's impossible to find a famous dataset compatible with our training procedure.
However, I can try implementing a DummyMultiModalDataset.

**Review by kalharko — APPROVED** [review]

File: mfai/torch/lightning_modules/clip.py
Comment (colon3ltocard) [reply mfai/torch/lightning_modules/clip.py:1]: Wrapping a "famous" dataset with NamedTensors should be feasible in a few lines of code, I don't see this as a major blocking issue and it would illustrate how to do so. I can do it if you want @tourniert

File: mfai/tokenizers.py [hunk 6]
```diff
@@ -108,28 +108,43 @@ def post_init(self, tokens: set) -> None:
...
-        # at this stage the lookup table are not initialised
-        self.token_to_id: dict[int, int] | None = None
-        self.id_to_token: dict[int, int] | None = None
+        self.token_to_id: dict[int, int] = dict()
+        self.id_to_token: dict[int, int] = dict()
+
+        self.post_init(tokens)
+
+    @abstractmethod
+    def get_set_tokens(self) -> set:
```
Comment (colon3ltocard) [inline mfai/tokenizers.py:129]: looks a bit like java code, maybe make a **tokens** property returning a set ?

File: mfai/torch/dummy_dataset.py [hunk 7]
```diff
@@ -121,3 +128,147 @@ def test_dataloader(self) -> DataLoader:
...
     def predict_dataloader(self) -> DataLoader:
         return DataLoader(self.dummy_predict, self.batch_size, shuffle=False)
+
+
+##########################################################################################################
+##################################         MultiModal Dummy           ####################################
+##########################################################################################################
+
+
+class MultiModalDummyDataset(Dataset):
```
Comment (colon3ltocard) [inline mfai/torch/dummy_dataset.py:138]:
```suggestion
class DummyMultiModalDataset(Dataset):
```

File: mfai/torch/dummy_dataset.py [hunk 8]
```diff
@@ -121,3 +128,147 @@ def test_dataloader(self) -> DataLoader:
...
+            feature_names=[f"feature_{i}" for i in range(self.nb_input_channels)],
+        )
+        # Create the random output and input tokens
+        random_text_len = torch.randint(2, self.context_length, (1,))
+        output_text = torch.randint(self.vocab_size, (random_text_len,))
+        input_text = output_text[:-1]
+        return images, input_text, output_text
+
+
+class MultiModalDummyDataModule(DummyDataModule):
```
Comment (colon3ltocard) [inline mfai/torch/dummy_dataset.py:182]:
```suggestion
class DummyMultiModalDataModule(DummyDataModule):
```

**Comment by colon3ltocard** [comment]
@tourniert could you add the CLIP model to the README.md to be consistent with our other models/implementations ?

**Review by LBerth — APPROVED** [review]

File: mfai/tokenizers.py [hunk 9]
```diff
@@ -108,28 +108,43 @@ def post_init(self, tokens: set) -> None:
...
-    narrow vocab problems with only 1000 tokens out of a vocab of 50000.
+    A Tokenizer using a reduced set of tokens from a base/parent Tokenizer.
+    Typical use case is for narrow vocab problems with only 1000 tokens out
+    of a vocab of 50000.
+    To use these class, you only have to implement the method 'get_set_tokens'.
     """
 
     def __init__(self, base_tokenizer: Tokenizer):
         self.base_tokenizer = base_tokenizer
+        tokens = self.tokens()
```
Comment (kalharko) [inline mfai/tokenizers.py:121]: What is the use of the line 121 `tokens = self.tokens()` ?

File: mfai/tokenizers.py [hunk 10]
```diff
@@ -108,28 +108,43 @@ def post_init(self, tokens: set) -> None:
...
-        # at this stage the lookup table are not initialised
-        self.token_to_id: dict[int, int] | None = None
-        self.id_to_token: dict[int, int] | None = None
+        self.token_to_id: dict[int, int] = dict()
+        self.id_to_token: dict[int, int] = dict()
+
+        self.post_init(tokens)
+
+    @abstractmethod
+    def tokens(self) -> set:
```
Comment (kalharko) [inline mfai/tokenizers.py:129]: Use of this function unclear. The exemple shows that the function returns a set of encoded strings, which are hard coded in the function ?
Is this used for testing ?

File: mfai/torch/dummy_dataset.py [hunk 11]
```diff
@@ -121,3 +129,147 @@ def test_dataloader(self) -> DataLoader:
...
+            "test",
+            self.dim_x,
+            self.dim_y,
+            self.nb_input_channels,
+            self.context_length,
+        )
+
+        self.eot_token = self.dummy_train.eot_token
+
+    def collate_fn_fit(
```
Comment (kalharko) [inline mfai/torch/dummy_dataset.py:229]: Excessive use of `# type: ignore` comments.
Proposition for better type hinted version of this function, wich uses only one ignore:
```py
    def collate_fn_fit(
        self,
        batch: List[Tuple[NamedTensor, torch.Tensor, torch.Tensor]]
    ) -> Tuple[NamedTensor, torch.Tensor, torch.Tensor]:
        """Collate a batch of multimodal data."""
        images: list[NamedTensor]
        input_txt: list[torch.Tensor]
        targets: list[torch.Tensor]
        images, input_txt, targets = zip(*batch)  # type: ignore[assignment]

        return (
            NamedTensor.collate_fn(images),
            self.collate_text(input_txt),
            self.collate_text(targets)
        )
     ```

File: mfai/tokenizers.py (hunk 9 above)
Comment (tourniert) [reply mfai/tokenizers.py:121]: self.tokens() is a method that return the set of all encoded words the MiniTokenizer has seen.
I move the call to self.tokens directly in post_init() for a simpler implementation.

File: mfai/tokenizers.py (hunk 10 above)
Comment (tourniert) [reply mfai/tokenizers.py:129]: The texts to be encoded are not necessarily hard-coded, on the contrary it is often necessary to use a third-party function to load them.
In the case of a unit test, however, I hardcoded the string to encode.

**Review by kalharko — APPROVED** [review]

**Review by colon3ltocard — APPROVED** [review]
